tuning.py

```diff
@@ -0,0 +1,317 @@
+    """
+    wt_seq = landscape.wildtype_sequence
+    all_singles = sampling.get_all_single_mutants(wt_seq, landscape.vocab_size)
+    df = experiment.get_fitness_df(all_singles, landscape.evaluate, wt_seq)
+    if adaptive:
+        singles = df[df.fitness >= threshold].sequence
+    else:
+        singles = df[df.fitness < threshold].sequence
+
+    if len(singles) < 2:
+        raise ValueError(f'Invalid Landscape: fewer than 2 singles for threshold {threshold}')
+
+    doubles = []
+    single_a = []
+    single_b = []
+    for a, b in itertools.combinations(singles, 2):
+        # check that this is a true double
+        if utils.get_mutation_positions(a, wt_seq) == utils.get_mutation_positions(b, wt_seq):
+            continue
+        double = utils.add_seqs(a, b, wt_seq)[0]
+        doubles.append(double)
+        single_a.append(a)
+        single_b.append(b)
+
+    doubles_df = experiment.get_fitness_df(doubles, landscape.evaluate, wt_seq)
+
+    single_a_fitness = landscape.evaluate(np.vstack(single_a))
+    single_b_fitness = landscape.evaluate(np.vstack(single_b))
+    doubles_df['a_fitness'] = single_a_fitness
+    doubles_df['b_fitness'] = single_b_fitness
+
+    residual = doubles_df.fitness - doubles_df.a_fitness - doubles_df.b_fitness + landscape.evaluate(wt_seq)
+    doubles_df['residual'] = residual
+    return doubles_df
+
+
+def get_epistasis_stats(landscape: potts_model.PottsModel,
+                        threshold: float = 0, adaptive: bool = True) -> Tuple[float, float]:
+    """Returns statistics about epistasis for combinations of singles.
+
+    Mean epistasis effect size is defined as the average effect of epistasis when two single mutants
+    from the selected set are combined. The rate of reciprocal sign epistasis is defined as the fraction
+    of epistatic interactions that have effects opposing the effects of the constituent singles. (i.e. the rate
+    of negative epistasis for adaptive singles in combination).
+
+    Args:
+      landscape: The landscape.
+      threshold: The threshold fitness for selected singles.
+      adaptive: If True, thresholded singles will have fitness >= `threshold`. If False,
+        selected singles will have fitness < `threshold`.
+
+    Returns:
+      A Tuple of length 2 where element [0] is the mean epistasis effect size
+      and element [1] is the rate of reciprocal sign epistasis.
+    """
+    doubles_df = get_doubles_df(landscape, threshold, adaptive)
+    residual = doubles_df.residual
+
+    mean_epistasis_effect = np.mean(residual)
+    if adaptive:
+        rate_reciprocal_epistasis = (residual < 0).sum() / residual.shape[0]
+    else:
+        rate_reciprocal_epistasis = (residual > 0).sum() / residual.shape[0]
+    return mean_epistasis_effect, rate_reciprocal_epistasis
+
+
+def get_mean_single_effect(landscape: potts_model.PottsModel, threshold: float = 0, adaptive: bool = True) -> float:
```

Comment:
I recommend removing default values for all helper functions that aren't top-level functions. I've often had bugs by misusing helpers where I didn't realize there were constants that had been set as default values.

Reply:
I think this is wise